### PR TITLE
[graphql] make GrapheneRepository lazy load the repository

### PR DESCRIPTION
**Pull request by alangenfeld**
restructure `GrapheneRepository` to init with minimal information and load objects when they are needed 

## How I Tested These Changes

existing coverage

### CODE REVIEW

**Comment by alangenfeld** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25122?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#25122** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25122?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#25121** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25121?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @alangenfeld and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/dagster-graphql/dagster_graphql/schema/asset_graph.py [hunk 1]
```diff
@@ -1280,11 +1280,7 @@ def resolve_partitionDefinition(
         return None
 
     def resolve_repository(self, graphene_info: ResolveInfo) -> "GrapheneRepository":
-        return external.GrapheneRepository(
-            graphene_info.context,
-            graphene_info.context.get_repository(self._repository_selector),
-            graphene_info.context.get_code_location(self._repository_selector.location_name),
-        )
+        return external.GrapheneRepository(self._repository_handle)
```
Comment (gibsondan) [inline python_modules/dagster-graphql/dagster_graphql/schema/asset_graph.py:1283]: weird import structure (before and after this change)

File: python_modules/dagster-graphql/dagster_graphql/schema/external.py [hunk 2]
```diff
@@ -273,160 +271,172 @@ class Meta:
...
-                branch_workspace=workspace_context,
-                base_workspace=base_deployment_context,
+        self._batch_loader = None
+
+        super().__init__(name=handle.repository_name)
+
+    def get_repository(self, graphene_info: ResolveInfo) -> ExternalRepository:
+        return graphene_info.context.get_repository(self._handle.to_selector())
+
+    def get_batch_loader(self, graphene_info: ResolveInfo):
```
Comment (gibsondan) [inline python_modules/dagster-graphql/dagster_graphql/schema/external.py:288]: any urge to do any of the async Loader business here?

**Review by gibsondan — APPROVED** [review]
very exciting

File: python_modules/dagster-graphql/dagster_graphql/schema/external.py (hunk 2 above)
Comment (alangenfeld) [reply python_modules/dagster-graphql/dagster_graphql/schema/external.py:288]: yea but not with much urgency

**Pull request merged**